File: cores/typescript/src/joypad.ts
// NES joypad state for two standard controllers (port of cores/c/src/joypad.c).

export class Joypad {
    static readonly ControllerCount = 2;
    static readonly ButtonCount = 8;

    current: Uint8Array = new Uint8Array(Joypad.ControllerCount);
    strobe: boolean = false;
    shift: Uint8Array = new Uint8Array(Joypad.ControllerCount);
    counter: Uint8Array = new Uint8Array(Joypad.ControllerCount);

    setButton(controller: number, button: number, pressed: boolean): void {
        if (controller < 0 || controller >= Joypad.ControllerCount || button < 0 || button >= Joypad.ButtonCount) return;
        const mask = 1 << button;
        if (pressed) this.current[controller] |= mask;
        else this.current[controller] &= (~mask & 0xFF);
        if (this.strobe) this.shift[controller] = this.current[controller];
    }

    writeStrobe(value: number): void {
        const newStrobe = (value & 0x01) !== 0;
        if (this.strobe && !newStrobe) {
            for (let c = 0; c < Joypad.ControllerCount; ++c) {
                this.shift[c] = this.current[c];
                this.counter[c] = 0;
            }
        }
        this.strobe = newStrobe;
        if (this.strobe) {
            for (let c = 0; c < Joypad.ControllerCount; ++c)
                this.shift[c] = this.current[c];
        }
    }

    read(controller: number): number {
        if (controller < 0 || controller >= Joypad.ControllerCount) return 1;
        if (this.strobe) return this.current[controller] & 0x01;
        let bit: number;
        if (this.counter[controller] < Joypad.ButtonCount)
            bit = (this.shift[controller] >> this.counter[controller]) & 0x01;
        else
            bit = 1;
        if (this.counter[controller] < 0xFF) this.counter[controller]++;
        return bit;
    }

    clear(): void {
        for (let c = 0; c < Joypad.ControllerCount; ++c) {
            this.current[c] = 0;
            if (this.strobe) this.shift[c] = 0;
        }
    }
}

File: cores/csharp/src/Joypad.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// NES joypad state for two standard controllers (port of src/joypad.rs).
public sealed class Joypad
{
    public const int ControllerCount = 2;
    public const int ButtonCount = 8;

    // Button bit indices.
    public const int ButtonA = 0;
    public const int ButtonB = 1;
    public const int ButtonSelect = 2;
    public const int ButtonStart = 3;
    public const int ButtonUp = 4;
    public const int ButtonDown = 5;
    public const int ButtonLeft = 6;
    public const int ButtonRight = 7;

    public byte[] Current = new byte[ControllerCount];
    public bool Strobe;
    public byte[] Shift = new byte[ControllerCount];
    public byte[] Counter = new byte[ControllerCount];

    public Joypad()
    {
        // zeroed by default
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void SetButton(int controller, int button, bool pressed)
    {
        if (controller < 0 || controller >= ControllerCount || button < 0 || button >= ButtonCount)
            return;
        byte mask = (byte)(1u << button);
        if (pressed) Current[controller] |= mask;
        else Current[controller] &= (byte)~mask;
        if (Strobe) Shift[controller] = Current[controller];
    }

    public void WriteStrobe(byte value)
    {
        bool newStrobe = (value & 0x01) != 0;
        if (Strobe && !newStrobe)
        {
            for (int c = 0; c < ControllerCount; ++c)
            {
                Shift[c] = Current[c];
                Counter[c] = 0;
            }
        }
        Strobe = newStrobe;
        if (Strobe)
        {
            for (int c = 0; c < ControllerCount; ++c)
                Shift[c] = Current[c];
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte Read(int controller)
    {
        if (controller < 0 || controller >= ControllerCount)
            return 1;
        if (Strobe)
            return (byte)(Current[controller] & 0x01);
        byte bit;
        if (Counter[controller] < ButtonCount)
            bit = (byte)((Shift[controller] >> Counter[controller]) & 0x01);
        else
            bit = 1;
        if (Counter[controller] < 0xFF)
            Counter[controller]++;
        return bit;
    }

    public void Clear()
    {
        for (int c = 0; c < ControllerCount; ++c)
        {
            Current[c] = 0;
            if (Strobe) Shift[c] = 0;
        }
    }
}
